### PR TITLE
Issue55

### PR DESCRIPTION
**Pull request by technetos**

### CODE REVIEW

File: hatch/src/cli/commands/mod.rs [hunk 1]
```diff
@@ -8,6 +8,7 @@ mod ops;
 use project::Project;
 use clap::{ ArgMatches, App };
 
+
```
Comment (DannyPeck) [inline hatch/src/cli/commands/mod.rs]: Can we remove this line?

File: hatch/src/cli/commands/new.rs [hunk 2]
```diff
@@ -41,12 +41,12 @@ impl<'new> New {
...
 
-  fn construct_deps_string(&self, deps: &Vec<(String, String)>) -> String {
+  fn construct_deps_string(&self, deps: &Vec<Dependency>) -> String {
     if deps.is_empty() {
       String::new()
     } else {
-      String::from("deps:\n") + deps.iter().map(|&(ref url, ref name)| {
-        format!("  {}: {}\n", &name[..], &url[..])
+      String::from("deps:\n") + deps.iter().map(|d| {
+        format!("  {}: {}\n", d.name(), d.url())
```
Comment (DannyPeck) [inline hatch/src/cli/commands/new.rs:49]: what are the leading spaces for?

**Review by DannyPeck — DISMISSED** [review]
Looks good! Just a few comments.

File: hatch/src/cli/commands/new.rs (hunk 2 above)
Comment (technetos) [reply hatch/src/cli/commands/new.rs:49]: Compliance with YAML

Comment (DannyPeck) [reply hatch/src/cli/commands/new.rs:49]: Fair enough.

File: hatch/src/yaml.rs [hunk 3]
```diff
@@ -73,9 +73,8 @@ fn do_parse(yml_vec: Vec<Yaml>) -> HatchResult<Project> {
   if let Some(d) = yml_vec[0]["deps"].as_hash() {
     deps = d
       .iter()
-      .map(|(k, v)| (k.as_str(), v.as_str()))
-      .filter(|&(k, v)| k.is_some() && v.is_some())
-      .map(|(k, v)| Dependency::new(k.unwrap().to_owned(), v.unwrap().to_owned()))
+      .filter_map(|(_k, v)| v.as_str())
+      .map(|v| Dependency::new(v.to_owned()))
       .collect();
```
Comment (xmclark) [inline hatch/src/yaml.rs:56]: so is filter_map filtering based on whether the as_str returns Some(_)?

**Review by DannyPeck — DISMISSED** [review]

File: hatch/src/repo.rs [hunk 4]
```diff
@@ -14,8 +14,8 @@ pub fn hatchfile_path(base: &Path) -> PathBuf {
   base.join("Hatch.yml")
 }
 
-pub fn clone_dep(repo: &(String, String), path: &Path) {
-  Repository::clone(&repo.0, path.join(&repo.1));
+pub fn clone_dep(url: &str, name: &str, path: &Path) {
```
Comment (DannyPeck) [inline hatch/src/repo.rs]: Should this function just take a single path with name joined before hand?

File: hatch/src/repo.rs [hunk 5]
```diff
@@ -41,7 +41,7 @@ pub fn clone_project_deps(path: &Path,
 
   // Clone the dependencies specified on the command line
   user_defined_deps.iter().for_each(|dep| {
-    clone_dep(&dep.as_pair(), path);
+    clone_dep(&dep.url(), &dep.name(), path);
```
Comment (DannyPeck) [inline hatch/src/repo.rs]: See comment above about clone_dep.

File: hatch/src/repo.rs [hunk 6]
```diff
@@ -65,11 +65,11 @@ pub fn clone_project_deps(path: &Path,
 }
 
 fn clone_nested_project_deps(path: &Path, visited: &mut HashSet<String>) -> HatchResult<bool> {
-  let current_project = yaml::parse(hatchfile_path(path).as_path())?;
+  let current_project = read_project(path)?;
```
Comment (DannyPeck) [inline hatch/src/repo.rs]: In the code where I have been using the tasks, I have been only importing the tasks module and using the syntax: `task::read_project(path)` to give a better context about what read_project is.

File: hatch/src/repo.rs [hunk 7]
```diff
@@ -65,11 +65,11 @@ pub fn clone_project_deps(path: &Path,
...
   current_project.deps().iter().for_each(|dep| {
     // If we have already pulled in a dep, dont do it again
     if !visited.insert(current_project.name().to_owned()) {
-      clone_dep(&dep.as_pair(), &modules_path(path).as_path());
+      clone_dep(&dep.url(), &dep.name(), &modules_path(path).as_path());
```
Comment (DannyPeck) [inline hatch/src/repo.rs]: See comment above about clone_dep.

**Review by DannyPeck — CHANGES_REQUESTED** [review]

File: hatch/src/assets/tests/builder.rs [hunk 8]
```diff
@@ -1,7 +1,7 @@
-use assets::builder::Builder as AssetBuilder;
-use assets::ProjectAsset;
-use project::{ Project, ProjectKind, LibraryKind };
-use std::path::PathBuf;
+#[cfg(test)] use assets::builder::Builder as AssetBuilder;
```
Comment (DannyPeck) [inline hatch/src/assets/tests/builder.rs]: After reading this documentation: https://doc.rust-lang.org/book/second-edition/ch11-03-test-organization.html can this annotation simply be added next to the test module definition within the mod.rs file and not on every one of these lines? Or is there another reason they need to be on every use statement?

File: hatch/src/task.rs [hunk 9]
```diff
@@ -18,10 +18,9 @@ pub fn generate_assets(project: &Project) -> HatchResult<()> {
 }
 
 pub fn platform_type() -> PlatformKind {
-  let info = os_info::get();
-    match *info.os_type() {
-      Type::Macos => PlatformKind::MacOS,
-      Type::Windows => PlatformKind::Windows,
-      _ => PlatformKind::Linux
-    }
+  match *os_info::get().os_type() {
```
Comment (DannyPeck) [inline hatch/src/task.rs:21]: Sure :P.

File: hatch/src/repo.rs (hunk 4 above)
Comment (xmclark) [reply hatch/src/repo.rs]: That would be intuitive. But I don't care that much.

**Review by xmclark — DISMISSED** [review]

File: hatch/src/yaml.rs (hunk 3 above)
Comment (technetos) [reply hatch/src/yaml.rs:56]: correct.

File: hatch/src/assets/tests/builder.rs (hunk 8 above)
Comment (technetos) [reply hatch/src/assets/tests/builder.rs]: You are absolutely correct!

**Review by DannyPeck — APPROVED** [review]
Looks good!

**Pull request merged**